### PR TITLE
Implement repository isolation, lazy download, and content handler app

### PR DESCRIPTION
**Pull request by gmbnomis**
- Change `CookbookPackageContent`'s natural key: Add random UUID for lazy download,
  sha256 for immediate download.

- Implement content app including '/universe' endpoint.  Remove previous
  live API.

- Introduce repo_key for `CookbookPackageContent`. This key must be unique within a repo version.
  For cookbooks, this is name & version

- Introduce stage that queries existing repo content and artifacts from the base
  repo version used for syncing (using repo key).  This implements repo
  isolation: content can only be shared within a repo unless it is the same
  (sha256 digest is known and is the same)

- Implement on_demand and streaming policies.

- Publish: Use repo key to check whether a repo version can be published without conflict.

### CODE REVIEW

**Comment by gmbnomis** [comment]
@mdellweg here is my take on repo isolation and lazy download policies for pulp_cookbook. Since pulp_gem seem to have similar challenges, I would love to get your feedback. It's a lot of stuff, but the most interesting part are the stages in `pulp_cookbook/app/tasks/synchronizing.py`, I think.

**Comment by mdellweg** [comment]
Need to take a closer look, but what is the reason for having a random uuid? Is it just for being unique?
Wouldn't it be enough, to use the RemoveDuplicate stage to implement the uniqueness per repository with less than a natural key?

 https://github.com/pulp/pulpcore-plugin/blob/15cfb26699f323c1d28951554e05be2d48470d63/pulpcore/plugin/stages/association_stages.py#L117

File: pulp_cookbook/app/tasks/synchronizing.py [hunk 1]
```diff
@@ -101,15 +293,22 @@ def __init__(self, remote):
...
-            ArtifactDownloader(), ArtifactSaver(),
-            ContentUnitSaver()
+            QueryExistingRepoContentAndArtifacts(new_version=new_version),
         ]
+        if self.download_artifacts:
+            pipeline.extend([
+                ArtifactDownloader(),
+                ArtifactSaver(),
+                UpdateContentWithDownloadResult(),
+                QueryExistingContentUnits(),
```
[the remark this answers is not in the document]
Comment (gmbnomis) [reply pulp_cookbook/app/tasks/synchronizing.py]: Good point. This uses the implementation of the stage within this file (although I stated above that I don't use it). It is used here in the "non-repo" mode of operation, i.e. it tries to identify content in the global content pool that is the same as the content just downloaded (i.e. name, version, sha256 in this case).

The stage acts like the standard  `QueryExistingContentUnits` stage with the exception that it ignores saved content (`unsaved_d_cs = [dc for dc in batch if dc.content.pk is None]`).

I think this change should go upstream in Pulp core. However, I am wondering whether there is use for the "look up in repo version only" mode of operation. This duplicates "QueryExistingRepoContentAndArtifacts" to some extent and I don't need it.

**Comment by gmbnomis** [comment]
> Need to take a closer look, but what is the reason for having a random uuid? Is it just for being unique?

I had a version with a non-unique 'natural key' that worked. Basically, it used name, version, and digest (sha256). digest could be null (lazy download case).

Content.q() queried for match in name & version & digest and digest != None. The problem with this approach is that it breaks too easily IMHO: We can't put in a DB uniqueness constraint on the Content and q() may return multiple matches. Some of the stage code can cope with such a q() other parts can't (but luckily, they can't be reached currently).

As all of the code in stages seems to be written with the underlying assumption that there is a unique natural key, I fear that a plugin implementation without such a key may break any time.

> Wouldn't it be enough, to use the RemoveDuplicate stage to implement the uniqueness per repository with less than a natural key?
> 
> https://github.com/pulp/pulpcore-plugin/blob/15cfb26699f323c1d28951554e05be2d48470d63/pulpcore/plugin/stages/association_stages.py#L117

I think that RemoveDuplicates does not work in this scenario, because there can't be duplicates: The first thing that happens in the pipeline is to connect (if possible) content units to existing saved content units & artifacts in the current repo version (using `QueryExistingRepoContentAndArtifacts`). If we find a match, we connect it to the existing unit in order not to have to download the corresponding artifact again.

But perhaps I don't understand your idea to use RemoveDuplicates for pulp_cookbook/pulp_gem?

(However, for other plugins, RemoveDuplicates could probably benefit from the `repo_q` and `repo_key` methods in Content: They could be the key used by default to remove duplicates.)

**Comment by mdellweg** [comment]
I am getting the feeling, that there is a lot to rethink around those natural keys.
It seems to defeat the whole purpose to have a digest in an natural key together with other fields.
On the other hand, considering content, we can never assume that anything but a digest is really unique across repositories.
Can we just drop the whole `natural_key` assumptions.
The `repo_key` might be the right solution to all this. After all it should not be to bad to duplicate content_units, as long as they share the artifacts.

**Comment by gmbnomis** [comment]
> I am getting the feeling, that there is a lot to rethink around those natural keys.

Agreed. As discussed, let's open the discussion in a Pulp issue. I am going to merge this now nevertheless. Because of the recent breaking changes in the plugin API, the old code simply does not run anymore and there is no point in adapting the old version.

**Pull request merged**